src/node/mod.rs

```diff
@@ -0,0 +1,382 @@
+//! Module for abstractions on drm device nodes.
+
+pub mod constants;
+
+use std::error::Error;
+use std::fmt::Debug;
+use std::fmt::{self, Display, Formatter};
+use std::io;
+use std::os::unix::io::AsFd;
+use std::path::{Path, PathBuf};
+
+use rustix::fs::{fstat, major, minor, stat, Dev as dev_t, Stat};
+
+use crate::node::constants::*;
+
+/// A node which refers to a DRM device.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct DrmNode {
+    dev: dev_t,
+    ty: NodeType,
+}
+
+impl DrmNode {
+    /// Creates a DRM node from an open drm device.
+    ///
+    /// This function does not take ownership of the passed in file descriptor.
+    pub fn from_file<A: AsFd>(file: A) -> Result<DrmNode, CreateDrmNodeError> {
+        let stat = fstat(file).map_err(Into::<io::Error>::into)?;
+        DrmNode::from_stat(stat)
+    }
+
+    /// Creates a DRM node from path.
+    pub fn from_path<A: AsRef<Path>>(path: A) -> Result<DrmNode, CreateDrmNodeError> {
+        let stat = stat(path.as_ref()).map_err(Into::<io::Error>::into)?;
+        DrmNode::from_stat(stat)
+    }
+
+    /// Creates a DRM node from a file stat.
+    pub fn from_stat(stat: Stat) -> Result<DrmNode, CreateDrmNodeError> {
+        let dev = stat.st_rdev;
+        DrmNode::from_dev_id(dev)
+    }
+
+    /// Creates a DRM node from a [`dev_t`].
+    pub fn from_dev_id(dev: dev_t) -> Result<DrmNode, CreateDrmNodeError> {
```

Comment:
Should we turn on `clippy::use_self`?

```suggestion
    pub fn from_dev_id(dev: dev_t) -> Result<Self, CreateDrmNodeError> {
```

Reply:
Honestly I think it's fine either way. I personally prefer `Self` because it makes text-based refactorings easier, but realistically most people are going to use LSP for stuff like that anyway so it rarely **really** matters.